### PR TITLE
Prepending batch dimensions to match Keras interfaces

### PR DESCRIPTION
**Pull request by cmdupuis3**
See issues #36 and #38

@jhamman @rabernat I'm not sure if this solves the problem in all cases, can you have a look?

### CODE REVIEW

**Comment by rabernat** [comment]
Thanks a lot for getting this PR started Chris!

> I'm not sure if this solves the problem in all cases, can you have a look?

It's not possible for me to know that just by looking at the code. That's why we have tests!

As you can [see](https://github.com/pangeo-data/xbatcher/runs/4230541155?check_suite_focus=true) your changes have caused existing tests to fail. That's because the generated batch dataset is being compared with an "expected" dataset.

https://github.com/pangeo-data/xbatcher/blob/d98ad21f65affa306dfd09d4b1d00e14f599fe65/xbatcher/tests/test_generators.py#L23-L32

Since the expected dataset does not have the extra dimension, the test fails. This raises the following question: _should this feature be optional?_ If so, we need to create a new option for it. Just changing the behavior of the code in this way is a "breaking change" to our API, and could cause problems for other xbatcher users. Tests help guard against this.

Changing the API is not out of the question--this is a very new and experimental package. But it would need to be discussed with the other developers and motivated by a clear argument.

If we do add an option for this feature (e.g. `squeeze_batch_dimension=True` as a default, but optionally `False` to enable the behavior in this PR), that needs to be covered by tests as well.

You can run the test suite yourself locally as

```python
pytest -v
```

(You'll need pytest installed.)

**Comment by cmdupuis3** [comment]
Sounds good, I'll try adding the squeeze option and see what happens. I'm guessing xbatcher is supporting non-Keras use cases as well?

**Comment by cmdupuis3** [comment]
Still need to add unit test(s)

**Comment by cmdupuis3** [comment]
@jhamman @rabernat I think this is ready to be merged

**Comment by cmdupuis3** [comment]
Doing some cleanup atm

**Comment by cmdupuis3** [comment]
@jhamman @rabernat Sorry for the confusion, I think this is in a mergeable state now.

File: xbatcher/generators.py [hunk 1]
```diff
@@ -105,6 +117,7 @@ def __init__(
         batch_dims={},
         concat_input_dims=False,
         preload_batch=True,
+        squeeze_batch_dim=True,
```
Comment (weiji14) [inline xbatcher/generators.py:120]: Is it ok to set the default to `True`? I.e. squeeze the batch dimension if all dims are included in the input dims. This might break existing behaviour, so wondering if the default should be `False` instead for backward compatibility.

**Review by weiji14 — COMMENTED** [review]
@maxrjones, since you're a bit more familiar with Keras, could you take a quick look at this PR in relation to #35 and #36 when you have time? This PR actually doesn't conflict too much with recent changes, though the tests might need some refactoring in line with #124.

That said, I'm thinking about how useful the `squeeze_batch_dim` is compared to just calling [`xr.DataArray.squeeze`](https://docs.xarray.dev/en/v2022.11.0/generated/xarray.DataArray.squeeze.html) or [`xr.DataArray.expand_dims`](https://docs.xarray.dev/en/v2022.11.0/generated/xarray.DataArray.expand_dims.html) manually. I suppose that adding this extra `squeeze_batch_dim` option increases usability, but it can also be a source of extra confusion, especially since the required shape of an input tensor is very subjective to what model architecture people are using.

File: xbatcher/generators.py (hunk 1 above)
Comment (cmdupuis3) [reply xbatcher/generators.py:120]: It would definitely break existing behavior. I'm not really sure which would be more intuitive, I think that's kind of subjective, but personally I would agree with you.

Comment (weiji14) [reply xbatcher/generators.py:120]: > The problem is that the existing behavior _already_ squeezes the array, so I think it would be kind of a pain to have to use `xr.DataArray.expand_dims`,

Yeah, I also find this unintuitive, the fact that xbatcher squeezes any non `input_dims` into a dim called `batch`. So another way to think of this `squeeze_batch_dim` feature is that xbatcher will just return the cropped/sliced/chipped array without squeezing the dims. This is more important with higher dimensional (＞3D) arrays because sometimes you do want to reserve the extra dims.

File: xbatcher/generators.py [hunk 2]
```diff
@@ -90,6 +98,10 @@ class BatchGenerator:
     preload_batch : bool, optional
         If ``True``, each batch will be loaded into memory before reshaping /
         processing, triggering any dask arrays to be computed.
+    squeeze_batch_dim : bool, optional
+        If ``False`` and all dims are input dims, each batch's dataset will have a
+        "batch" dimension of size 1 prepended to the array. This functionality is
```
Comment (weiji14) [inline xbatcher/generators.py:103]: Some suggestions on the documentation. Maybe L86 and L87 could be edited to indicate the squeeze behavior is controllable with `squeeze_batch_dim`?

Also, this sentence might be a bit confusing to read. So an extra dimension of size 1 is added/prepended only when `batch_dims is None` or unset. For cases where `len(batch_dims) ＞= 1)`, the squeezing/collapsing of dimensions is still happening though. I'm wondering what's a good way to reword this to make it clearer on what is happening and why this option exists.

Comment (cmdupuis3) [reply xbatcher/generators.py:103]: The problem would only appear in this one corner case, as far as I could tell. So, my solution was coded to only apply when there were no `batch_dims` (at least that was my intention). If you're changing the default behavior though, you can probably make this simpler, and therefore the docs would be less confusing too.

**Comment by cmdupuis3** [comment]
Actually, if we go with the proposal of default `False` and rely on `xr.DataArray.squeeze`, this option becomes completely unnecessary.

**Comment by weiji14** [comment]
> Actually, if we go with the proposal of default `False` and rely on `xr.DataArray.squeeze`, this option becomes completely unnecessary.

Hmm, let's get a second opinion from someone who uses Keras. If they agree, we can close this Pull Request?

**Comment by maxrjones** [comment]
Sorry for taking so long to respond here. My understanding of the conversation above is that the purpose of this PR can be broken down to two issues:

1. Should there be an easier option to retain all the original dataset dimensions in the generated batches? Currently one would need to include all the original dims in the dict provided to `input_dims`. My opinion is that this feature would be useful. One option would be for the user to specify `input_dims=False` to avoid any stacking step. Since this is not the main purpose of this PR, I suggest opening a separate issue to discuss further if any of you agree that this is worthwhile.
2. Should there always be a `sample` dimension (previously called the `batch` dimension)? This is the topic of https://github.com/xarray-contrib/xbatcher/issues/127. If we chose to implement the feature suggested in point 1, I think it makes sense to include a `sample` dimension in the case in which the user specifies that stacking should take place. In the case without a `.stack` step, I think it’s trivial to add a length 1 dimension and that it is not worth the API complexity to support that through the batch generator.

If either of these points would benefit from some in-depth discussion, we could put this PR on the agenda for next Monday’s Pangeo ML working group meeting.

**Comment by cmdupuis3** [comment]
Hi all, just wondering what the status of this is? 

I'm writing some xbatcher example notebooks and I'm running into this issue again.

**Comment by maxrjones** [comment]
> Hi all, just wondering what the status of this is?
> 
> I'm writing some xbatcher example notebooks and I'm running into this issue again.

Apologies again for the delay - my attention was drawn elsewhere due to illness, travel, and some other urgent tasks. If I recall correctly we proposed at our last meeting to always have a defined axis order including a sample dimension and provide the option to specify that no transposing/stacking should occur using `input_dims=None`. I will try to draft this alternative solution by Monday.
